File: lintcode/binary_tree_maximum_path_sum.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        def helper(root):
            if not root: return (None, 0)
            (l_max, l) = helper(root.left)
            (r_max, r) = helper(root.right)
            return (max(filter(lambda x: x is not None, [l_max, r_max, 
                    l + r + root.val])), max(max(l, r) + root.val, 0))
        return helper(root)[0]

File: leetcode/binary_tree_maximum_path_sum.js
/* Given a binary tree, find the maximum path sum.  
 * The path may start and end at any node in the tree.
 * Time Complexity: O(n)
 * Space Complexity: O(h), h is the height of the tree
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    function helper(root) {
        if (!root)
            return [null, 0];
        var left = helper(root.left);
        var right = helper(root.right);
        var first = left[0];
        if (!first || (right[0] && right[0] > first))
            first = right[0];
        if (!first || (left[1] + right[1] + root.val > first))
            first = left[1] + right[1] + root.val;
        return [first, Math.max(0, root.val + Math.max(left[1], right[1]))];
    }
    return helper(root)[0];
};
